compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -289,7 +291,7 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
 
         // For simplicities' sake, we implement reallocate as "alloc, copy, dealloc".
         // This happens so rarely, the perf advantage is outweighed by the maintenance cost.
-        let new_ptr = self.allocate_ptr(new_size, new_align, kind)?;
+        let new_ptr = self.allocate_ptr(new_size, new_align, kind, zero_init)?;
```

Comment:
Since we'll memcopy right over the alloc, zeroing is not needed for realloc, just use false here

Reply:
oh I see, thanks!

Reply:
Please add a comment explaining this.

Reply:
Also seems worth saying that we assume here that zeroing the entire allocation is more efficient than zeroing just the parts not copied from the old allocation.

Reply:
I don't understand this reply, sorry. A comment should still be added.

Reply:
> creating an uninitialized allocation and a zero-initialized allocation takes the same amount of work, memory and implementation effort

I don't think that's true. The allocator can rely on the fact that `mmap`ed pages are zeroed, but the allocator will likely also recycle storage without giving it back to the OS in the mean time, and then it has to be explicitly reset to 0.

Reply:
We explicitly call `Bytes::allocate_zeroed` because we just want a big `Box<[u8]>` and not have to handle uninit bytes for the allocation bytes

Reply:
So yes, we can rely on this. This PR doesn't actually change how memory is initialized, only what is set in the init mask

Reply:
Oh, right.